tensorflow_addons/optimizers/discriminative_layer_training_test.py

```diff
@@ -0,0 +1,467 @@
+
+        learning_rate = 0.01
+        model_lr = model_fn()
+        model_lr.layers[0].lr_mult = 0.3
+        model_lr.layers[0].layers[-1].lr_mult = 0.1
+        model_lr.layers[-1].lr_mult = 0.5
+
+        d_opt = DiscriminativeLayerOptimizer(
+            opt, model_lr, verbose=False, learning_rate=learning_rate
+        )
+        model_lr.compile(loss=loss, optimizer=d_opt)
+
+        x = np.ones(shape=(8, 32, 32, 3), dtype=np.float32)
+        y = np.zeros(shape=(8, 5), dtype=np.float32)
+        y[:, 0] = 1.0
+
+        filepath = os.path.join(tempfile.gettempdir(), model_fn.__name__ + "_cp.ckpt")
+
+        callbacks = [
+            tf.keras.callbacks.ModelCheckpoint(
+                filepath=filepath, save_weights_only=True, verbose=1
+            )
+        ]
+
+        model_lr.fit(
+            x,
+            y,
+            epochs=2,
+            batch_size=4,
+            verbose=False,
+            shuffle=False,
+            callbacks=callbacks,
+        )
+
+        # If this doesn't error out, then loading and checkpointing should be fine.
+        model_lr.load_weights(filepath=filepath)
+
+    def _test_config_tofrom(self, model_fn, loss, opt):
+        """Confirm that optimizer saves config and loads config."""
+
+        # build model and save the opt to a config as c.
+        learning_rate = 0.01
+        model_lr = model_fn()
+        model_lr.layers[0].lr_mult = 0.3
+        model_lr.layers[0].layers[-1].lr_mult = 0.1
+        model_lr.layers[-1].lr_mult = 0.5
+
+        d_opt = DiscriminativeLayerOptimizer(
+            opt, model_lr, verbose=False, learning_rate=learning_rate
+        )
+        model_lr.compile(loss=loss, optimizer=d_opt)
+
+        c = d_opt.get_config()
+
+        # reconstruct the model and then build the opt from config.
+
+        model_lr = model_fn()
+        model_lr.layers[0].lr_mult = 0.3
+        model_lr.layers[0].layers[-1].lr_mult = 0.1
+        model_lr.layers[-1].lr_mult = 0.5
+
+        d_opt_from_config = DiscriminativeLayerOptimizer.from_config(c, model_lr)
+        model_lr.compile(loss=loss, optimizer=d_opt_from_config)
+
+        # we expect both optimizers to have the same optimizer group and base optimizer.
+        self.assertAllEqual(
+            len(d_opt.optimizer_group), len(d_opt_from_config.optimizer_group)
+        )
+        self.assertAllEqual(d_opt.opt_class, d_opt_from_config.opt_class)
+
+        # we also expect the lr for each opt in the opt groups to be the same. Also confirms same lr mult.
+        self.assertAllEqual(
+            [opt.learning_rate for opt in d_opt.optimizer_group],
+            [opt.learning_rate for opt in d_opt_from_config.optimizer_group],
+        )
+
+
+def wrap_test(func, **kwargs):
+    """Wrap the test method so that it has pre assigned kwargs."""
+
+    def test(self):
+        return func(self, **kwargs)
+
+    return test
+
+
+def generate_tests():
```

Comment:
please use `pytest.mark.parametrize` instead. It works with arbitrary python object and should end up much simpler.